### PR TITLE
feat: Implement more efficient version of xxhash64

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion-comet/issues/547

## Rationale for this change

The twox-hash crate seems to be designed for use cases such as processing files where it is important to process data in chunks and not need to load data fully into memory first. In our use case, our data is already in memory, so we can optimize for that.

### Before

Running on MacBook Pro M3 Max.

```
hash/xxhash64/8192      time:   [892.53 µs 893.12 µs 893.85 µs]
```

### After

```
hash/xxhash64/8192      time:   [209.25 µs 209.50 µs 209.78 µs]
                        change: [-76.607% -76.579% -76.552%] (p = 0.00 < 0.05)
                        Performance has improved.
```

## What changes are included in this PR?

- Implement simpler version of xxhash64

### CODE REVIEW

**Comment by andygrove** [comment]
@parthchandra @advancedxy This is ready for review now

File: NOTICE.txt
Comment (kazuyukitanimura) [inline NOTICE.txt:3]: hmmm so does this mean that comet would be dual licensed?

Not sure the legal part... Especially the Copyright github url part...

Comment (andygrove) [reply NOTICE.txt:3]: It would be good to get another opinion on this though. Perhaps @alamb could offer some thoughts.

File: core/src/execution/datafusion/spark_hash.rs
Comment (andygrove) [reply core/src/execution/datafusion/spark_hash.rs]: Thanks for the review @advancedxy. I moved xxhash64 into it's own file under `execution.datafusion.expressions` which makes sense also because this is a regular SQL function that users can call.

**Review by alamb — COMMENTED** [review]
I think deriving code from MIT licensed code is fine, and consistent with how the Arrow project does things: https://github.com/apache/arrow/blob/main/LICENSE.txt

FYI @shepmaster

**Comment by andygrove** [comment]
@parthchandra @kazuyukitanimura could I get a commiter review?

**Review by viirya — APPROVED** [review]
I read it and compare with https://github.com/shepmaster/twox-hash/blob/5f3f3a345be5f65680c5f2b9ed5950c85e9a0ccf/src/sixty_four.rs#L27. Looks the same logic. The random comparison test is also passed.

**Comment by viirya** [comment]
The improvement is nice. 💯

**Review by kazuyukitanimura — APPROVED** [review]
Just some comments on license.

File: LICENSE.txt
Comment (andygrove) [reply LICENSE.txt:219]: Thanks, I will add this

Comment (andygrove) [reply LICENSE.txt:219]: Added

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (parthchandra) [inline core/src/execution/datafusion/expressions/xxhash64.rs:148]: I know this is based on `twox-hash` which also does what you're doing here. However, this looks like this is the [XXH64_mergeRound](https://github.com/Cyan4973/xxHash/blob/805c00b68fa754200ada0c207ffeaa7a4409377c/xxhash.h#L3422) function which calls [XXH64_Round](https://github.com/Cyan4973/xxHash/blob/805c00b68fa754200ada0c207ffeaa7a4409377c/xxhash.h#L3397) which is slightly different.
XXH64_Round seems to be `ingest_one_number`

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (parthchandra) [reply core/src/execution/datafusion/expressions/xxhash64.rs:141]: Oh, nvm. I was reading this wrong.

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (parthchandra) [inline core/src/execution/datafusion/expressions/xxhash64.rs:22]: Just curious, is this the preferred way to write these or would hex be preferred (one is just as unreadable as the other).

**Review by parthchandra — APPROVED** [review]
This looks good to me.
I am not sure we have exactly the same implementation as the original (or the same as Spark, for that matter). If the fuzz tests pass I have no concerns.

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (andygrove) [reply core/src/execution/datafusion/expressions/xxhash64.rs:22]: :man_shrugging: I copied this directly from xxhash64. I think either style works in this case

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (andygrove) [reply core/src/execution/datafusion/expressions/xxhash64.rs:46]: This is also processing 64 bits at a time. We are calling `read_unaligned()` on a `ptr_u64`.

File: core/src/execution/datafusion/expressions/xxhash64.rs
Comment (andygrove) [reply core/src/execution/datafusion/expressions/xxhash64.rs:148]: I think the logic looks the same?

XXH64_mergeRound calles XXH64_Round which does this:

```
acc += input * XXH_PRIME64_2;
acc  = XXH_rotl64(acc, 31);
acc *= XXH_PRIME64_1;
```

Then XXH64_mergeRound does this:
    
```
acc ^= val;
acc  = acc * XXH_PRIME64_1 + XXH_PRIME64_4;
```

**Comment by shepmaster** [comment]
Are your benchmarks available for me to try out? I actually reimplemented the algorithm from scratch to see if there were any changes I would make in today's Rust. When I benchmarked it, I didn't have much difference compared to my existing implementation, but the numbers do noticeably differ from yours.

> ### Before
> Running on MacBook Pro M3 Max.
> 
> ```
> hash/xxhash64/8192      time:   [892.53 µs 893.12 µs 893.85 µs]
> ```

My guess is that this is hashing a buffer of 8192 bytes, but that would result in ~9 MiB/s which would be really slow. Maybe it's some other unit?

I grabbed the implementation code from this PR and popped it into my benchmark and got:

```
head-to-head/twox-hash/8192
                        time:   [565.59 ns 567.14 ns 569.04 ns]
                        thrpt:  [13.408 GiB/s 13.452 GiB/s 13.489 GiB/s]
Found 8 outliers among 100 measurements (8.00%)
  3 (3.00%) high mild
  5 (5.00%) high severe

head-to-head/spark/8192 time:   [565.39 ns 566.90 ns 568.63 ns]
                        thrpt:  [13.417 GiB/s 13.458 GiB/s 13.494 GiB/s]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
```

This is on a MacBook Pro M1 Max, so if anything I'd expect it to be slower than your test machine.

<details>
<summary>Benchmark code</summary>

```rust
use criterion::{criterion_group, criterion_main, Criterion, Throughput};
use rand::{Rng, RngCore, SeedableRng};
use std::hash::Hasher;
use std::hint::black_box;
use twox_hash::XxHash64;

fn head_to_head(c: &mut Criterion) {
    let (seed, data) = gen_data();
    let mut g = c.benchmark_group("head-to-head");

    for size in [8192] {
        let data = &data[..size];
        g.throughput(Throughput::Bytes(data.len() as _));

        let id = format!("twox-hash/{size}");
        g.bench_function(id, |b| {
            b.iter(|| {
                let hash = {
                    let mut hasher = XxHash64::with_seed(seed);
                    hasher.write(&data);
                    hasher.finish()
                };
                black_box(hash);
            })
        });

        let id = format!("spark/{size}");
        g.bench_function(id, |b| {
            b.iter(|| {
                let hash = extracto::spark_compatible_xxhash64(&data, seed);
                black_box(hash);
            })
        });
    }

    g.finish();
}

const SEED: u64 = 0xc651_4843_1995_363f;
const DATA_SIZE: usize = 100 * 1024 * 1024;

fn gen_data() -> (u64, Vec<u8>) {
    let mut rng = rand::rngs::StdRng::seed_from_u64(SEED);

    let seed = rng.gen();

    let mut data = vec![0; DATA_SIZE];
    rng.fill_bytes(&mut data);

    (seed, data)
}

criterion_group!(benches, head_to_head);
criterion_main!(benches);
```

</details>

**Comment by andygrove** [comment]
Thanks for spending time looking at this @shepmaster. I just ran the benchmarks again using the following steps and still see the same results I initially posted here.

1. checkout andygrove:xxhash64-stateless (this PR)
2. run `cargo bench xxhash64`
3. take backup of `core/benches/hash.rs`
4. checkout main branch
5. restore backup of `core/benches/hash.rs`
6. run `cargo bench xxhash64` 

Results from step 2:

```
hash/xxhash64/8192      time:   [210.40 µs 210.55 µs 210.72 µs]
                        change: [-0.1589% -0.0179% +0.1328%] (p = 0.81 > 0.05)
                        No change in performance detected.
```

Results from step 6:

```
hash/xxhash64/8192      time:   [891.98 µs 894.49 µs 898.04 µs]
                        change: [+323.31% +324.24% +325.23%] (p = 0.00 < 0.05)
                        Performance has regressed.
```

**Comment by shepmaster** [comment]
Thank you! I was able to run the benchmarks and tweak my local version of the code to match the performance of your PR. I'll work on rolling that into twox-hash proper and release a new version at some point. Thanks for the fresh perspective on how to improve performance!

**Comment by andygrove** [comment]
Thanks for all the reviews @viirya @parthchandra @kazuyukitanimura @advancedxy. I will go ahead and merge this now but let me know if you have any more comments or questions.

**Pull request merged**
